### PR TITLE
advanced-text-layer

### PR DESCRIPTION
**Pull request by georgios-uber**
This is still WORK-IN-PROGRESS

### CODE REVIEW

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+
```
Comment (1chandu) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.md]: Can you describe the difference between `TextLayer` and `AdvancedTextLayer`.

**Comment by georgios-uber** [comment]
@1chandu It will render using distance fields, support distance fields effects (like glow) and position characters based on font info and kerning.

**Comment by jianhuang01** [comment]
@georgios-uber Distance fields effects & kerning are necessary for all texts, can we merge these effects to current textLayer instead of creating a new layer?

**Comment by georgios-uber** [comment]
@jianhuang01 I want to develop this without worrying of breaking existing functionality.
Eventually we can merge them.

**Comment by jianhuang01** [comment]
@georgios-uber Thanks, that's good.

**Comment by georgios-uber** [comment]
screen shot 2018-03-26 at 7 28 00 pm

**Comment by georgios-uber** [comment]
@ibgreen I got the calculations to work properly. Please take a look at the updated code.

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+          changeFlags.updateTriggersChanged.getText ||
+          changeFlags.updateTriggersChanged.getPosition))
+    ) {
+      this.transformStringToLetters();
+    }
+  }
+
+  transformStringToLetters() {
+    const {data, getText, getPosition} = this.props;
+    if (!data || data.length === 0) {
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: empty data is always forced to empty array, you should not need this check.

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+          const x = xpos + (width / 2.0 - xoffset);
+          const y = (height - base) / 2.0 + yoffset;
+          xpos += xadvance;
+          // console.log(letter, {xoffset, yoffset, width, height, base});
+
+          return ({icon: letter, position, x, y});
+        }
+        );
+      })
+      .reduce((prev, curr) => [...prev, ...curr]);
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Maybe add a comment. I am not 100% sure how this works. Doesn't reduce need a start value?

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    if (!data || data.length === 0) {
+      return;
+    }
+
+    const {iconMapping} = this.state;
+    if (!iconMapping) {
+      return;
+    }
+
+    const transformedData = data
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: This function seems a little inefficient. But OK for now.

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        }
+        );
+      })
+      .reduce((prev, curr) => [...prev, ...curr]);
+
+    this.setState({data: transformedData});
+  }
+
+  getAnchorXFromTextAnchor(textAnchor) {
+    if (!TEXT_ANCHOR.hasOwnProperty(textAnchor)) {
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Are checking and error strings worth the bundle size cost? Or just do
` return TEXT_ANCHOR[textAnchor] || 0;`

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+# Text Layer
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.md]: Advanced Text Layer?

File: src/experimental-layers/src/advanced-text-layer/text-multi-icon-layer.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import {IconLayer} from 'deck.gl';
+import vs from './advanced-text-layer-vertex.glsl';
+import fs from './advanced-text-layer-fragment.glsl';
+
+
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/text-multi-icon-layer.js]: Don't forget to declare props.

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+  getText: x => x.text,
+  getPosition: x => x.coordinates,
+  getColor: x => x.color || DEFAULT_COLOR,
+  getSize: x => x.size || 32,
+  getAngle: x => x.angle || 0,
+  getTextAnchor: x => x.textAnchor || 'middle',
+  getAlignmentBaseline: x => x.alignmentBaseline || 'center',
+  getPixelOffset: x => x.pixelOffset || [0, 0],
+  fp64: false
+};
```
Comment (ibgreen) [inline src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: declare all props, `fontTexture, fontInfo, ...`. Set them to `null` if no value.

**Comment by ibgreen** [comment]
> I got the calculations to work properly. Please take a look at the updated code.

@georgios-uber that looks super smooth!!!

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer-fragment.glsl.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+varying float vColorMode;
+varying vec4 vColor;
+varying vec2 vTextureCoords;
+
+void main(void) {
+  float distance = texture2D(iconsTexture, vTextureCoords).a;
+  float alpha = smoothstep(0.5 - smoothing, 0.5 + smoothing, distance);
+  gl_FragColor = vec4(vColor.rgb, vColor.a * alpha * opacity);
+
+  // float outlineFactor = smoothstep(0.5 - smoothing, 0.5 + smoothing, distance);
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer-fragment.glsl.js]: OK, will cut

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+attribute vec2 instanceOffsets;
+
+uniform float sizeScale;
+uniform vec2 iconsTextureDim;
+
+varying float vColorMode;
+varying vec4 vColor;
+varying vec2 vTextureCoords;
+
+vec2 rotate_by_angle(vec2 vertex, float angle) {
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js]: I copied this from icon layer. Maybe we should fix that instead?

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  float cos_angle = cos(angle_radian);
+  float sin_angle = sin(angle_radian);
+  mat2 rotationMatrix = mat2(cos_angle, -sin_angle, sin_angle, cos_angle);
+  return rotationMatrix * vertex;
+}
+
+void main(void) {
+  vec2 iconSize = instanceIconFrames.zw;
+
+  // scale and rotate vertex in "pixel" value and convert back to fraction in clipspace
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js]: I copied this from icon layer. Maybe we should fix that instead?

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  vec2 pixelOffset = positions / 2.0 * iconSize + instanceOffsets;
+  pixelOffset = rotate_by_angle(pixelOffset, instanceAngles) * sizeScale * instanceSizes;
+  pixelOffset.y *= -1.0;
+
+  gl_Position = project_position_to_clipspace(instancePositions, instancePositions64xyLow, vec3(0.0));
+  gl_Position += project_pixel_to_clipspace(pixelOffset);
+
+  vTextureCoords = mix(
+    instanceIconFrames.xy,
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js]: I copied this from icon layer. Maybe we should fix that instead?

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    instanceIconFrames.xy + iconSize,
+    (positions.xy + 1.0) / 2.0
+  ) / iconsTextureDim;
+
+  vTextureCoords.y = 1.0 - vTextureCoords.y;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer-vertex.glsl.js]: I copied this from icon layer. Maybe we should fix that instead?

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js (hunk 8 above)
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: OK

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+          };
+        });
+        this.setState({iconMapping});
+      });
+    });
+  }
+
+  updateState({props, oldProps, changeFlags}) {
+    if (
+      changeFlags.dataChanged ||
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Again this is copied. We do it in most layers this exact way...

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js (hunk 3 above)
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Again copied from text layer.

Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Maybe should rewrite

File: src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js (hunk 5 above)
Comment (georgios-uber) [reply src/experimental-layers/src/advanced-text-layer/advanced-text-layer.js]: Again copied code, right now anchors are not supported/functional...

**Pull request merged**